File: editor/en/[988]Smallest String Starting From Leaf.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    String smallest = null;

    public String smallestFromLeaf(TreeNode root) {
        dfs(root, new StringBuilder());
        return smallest;
    }

    private void dfs(TreeNode node, StringBuilder sb) {
        if (node == null) {
            return;
        }

        sb.append((char) (node.val + 'a'));

        if (node.left == null && node.right == null) {
            String leafToRoot = sb.reverse().toString();
            sb.reverse();

            if (smallest == null || leafToRoot.compareTo(smallest) < 0) {
                smallest = leafToRoot;
            }
        }

        dfs(node.left, sb);
        dfs(node.right, sb);
        sb.deleteCharAt(sb.length() - 1);
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[988]Smallest String Starting From Leaf.py
# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.smallest = None

    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        self.dfs(root, [])
        return self.smallest

    def dfs(self, node, path) -> None:
        if node is None:
            return

        path.append(chr(node.val + ord('a')))

        if node.left is None and node.right is None:
            leaf_to_root = ''.join(reversed(path))
            if self.smallest is None or leaf_to_root < self.smallest:
                self.smallest = leaf_to_root

        self.dfs(node.left, path)
        self.dfs(node.right, path)
        path.pop()
# leetcode submit region end(Prohibit modification and deletion)
